### ISSUE TITLE
Nonstandard YAML config format results in crash

### ISSUE THREAD

**Issue by naoyak** [airbnb/omniduct#97]
In Omniduct 1.x, a YAML format Duct config with inconsistent depths can cause `Duct.__init__` to crash.

Format is something like this:

```yaml
data_environments:
    env1:
        databases:
            presto:
                protocol: presto
                host: localhost
                port: 9999
                catalog: silver
                schema: default
            hive:
                protocol: hiveserver2
                host: localhost
                port: 3333
                schema: default
                driver: pyhive
        metastore:
            host: localhost
            port: 5555

     env2:
        databases:
            presto:
                protocol: presto
                host: localhost
                port: 6666
                catalog: gold
                schema: default
            hive:
                protocol: hiveserver2
                host: localhost
                port: 2222
                cache: local
                schema: default
                driver: pyhive
        metastore:
            host: localhost
            port: 8888

default_data_environment: env1
```

Error:
```
/efs/home/naoya_kanai/repos/omniduct/omniduct/registry.py in __init__(self, config)
     56         if config:
     57             print('init: ', config)
---> 58             self.register_from_config(config)
     59 
     60     def __repr__(self):

/efs/home/naoya_kanai/repos/omniduct/omniduct/registry.py in register_from_config(self, config)
    256             names = duct_config.pop('name')
    257             print(names)
--> 258             protocol = duct_config.pop('protocol')
    259             register_magics = duct_config.pop('register_magics', True)
    260             try:

KeyError: 'protocol'
```

**Comment by matthewwardrop** [comment]
Roger that... I'll look into this tonight. Thanks for the report!

**Comment by naoyak** [comment]
Thanks @matthewwardrop! For context, I think this is happening because the envs defined in the YAML config have a mix of `Duct`s (Hive, Presto) and unrelated stuff which doesn't follow the convention expected by Omniduct. I suppose this is something that could be handled gracefully, e.g. by ignoring sections which don't look like what's expected.

**Comment by matthewwardrop** [comment]
@naoyak Hmmm... looking at this more closely, this failure was "designed", in that it didn't seem sensible to me at the time to have inconsistently nested definitions, which could create ambiguous definitions. In this case, why not just suppress the `database` nesting, and instead have:
```
data_environments:
    env1:
        presto:
            protocol: presto
            host: localhost
            port: 9999
            catalog: silver
            schema: default
        hive:
            protocol: hiveserver2
            host: localhost
            port: 3333
            schema: default
            driver: pyhive
        metastore:
            host: localhost
            port: 5555
```

If `metastore` is not a duct, then just create a new nesting for `omniduct` or `services`, and pass only the subset of configurations to Omniduct, like:
```
data_environments:
    env1:
        services:
             databases: ....
        metastore:
             ....
````

**Comment by naoyak** [comment]
@matthewwardrop thanks for checking it out! Yeah, you're probably right about this. It seems to be as simple as either tweaking the YAML format or tweaking the read-in logic, maybe even a one-liner. It's on the radar internally and should be fixed there.

**Comment by matthewwardrop** [comment]
Thanks for reporting in @naoyak.

**Comment by naoyak** [comment]
I'm going to reopen this for discussion - if we can provide a simple patch which handles "empty" nesting levels gracefully, would it be acceptable to re-enable the handling? What do you think @matthewwardrop?

Again for context, we have some production libraries which currently pin an old version of Omniduct in order to deal with a specific legacy YAML config on the servers. cc @oneraynyday

**Comment by matthewwardrop** [comment]
By "empty" do you mean levels which do not have a "protocol" field? I think I'd even be willing to change the parser to just check if there is a "protocol" field at any level, and if so, interpret that as a Duct configuration. Would that work for you?

**Comment by naoyak** [comment]
@matthewwardrop I think that would work! If Omniduct simply ignores entries w/o a `protocol` entry instead of throwing the exception, it would likely accommodate our usage.

@OneRaynyDay can you confirm?

**Comment by matthewwardrop** [comment]
@naoyak @OneRaynyDay I've included a fix for this in Omniduct 1.1.9 (https://github.com/airbnb/omniduct/commit/929a39a51e74bc9f6188e3e9b4d8ef9ef5557398). It's a more generous parser that traverses the configuration looking for any potential Duct configurations, and so allows configurations like the one you provided. It's also backward compatible with any configurations for Omniduct <= 1.1.8. I'll close this issue out; but feel free to reopen if this doesn't solve your issues.

**Comment by naoyak** [comment]
Thanks @matthewwardrop! This looks great, we'll go with 1.1.9 going forward and hopefully this issue will be resolved for good on our end.
